### PR TITLE
Add cal_logs to wfi_image-1.0.0

### PR DESCRIPTION
**Pull request by eslavich**
This adds an array of structured log information to wfi_image-1.0.0.  The intent is to use it to store log output from the romancal pipeline/step that produced the file.

### CODE REVIEW

File: src/rad/resources/schemas/cal_logs-1.0.0.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+%YAML 1.1
+---
+$schema: asdf://stsci.edu/datamodels/roman/schemas/rad_schema-1.0.0
+id: asdf://stsci.edu/datamodels/roman/schemas/cal_logs-1.0.0
+
+title: Calibration logs
+type: array
```
Comment (eslavich) [inline src/rad/resources/schemas/cal_logs-1.0.0.yaml:7]: Referendum: would it be better for this schema to describe a single log entry and not an array of logs?

**Comment by perrygreenfield** [comment]
Before approving, I'd like to see an example of how the log output actually looks like in the yaml header. Is there a good example available?

**Comment by eslavich** [comment]
Switching this to an array of strings instead of log message objects after a discussion with @perrygreenfield on Slack, stand by...

**Comment by eslavich** [comment]
Okay, this is ready to review again

**Comment by eslavich** [comment]
Sample ASDF:
```
cal_logs: !<asdf://stsci.edu/datamodels/roman/tags/cal_logs-1.0.0>
  - '2021-11-17T19:45:16.299Z :: stpipe.LoggingStep :: INFO :: Step LoggingStep running
    with args ().'
  - '2021-11-17T19:45:16.300Z :: stpipe.LoggingStep :: INFO :: Step LoggingStep parameters
    are: {''pre_hooks'': [], ''post_hooks'': [], ''output_file'': None, ''output_dir'':
    None, ''output_ext'': ''.asdf'', ''output_use_model'': False, ''output_use_index'':
    True, ''save_results'': False, ''skip'': False, ''suffix'': None, ''search_output_file'':
    True, ''input_dir'': ''''}'
  - '2021-11-17T19:45:16.300Z :: stpipe.LoggingStep :: WARNING :: Splines failed to
    reticulate'
```
The line wrapping is due to a pyyaml default setting that asdf currently does not provide access to.

**Comment by sosey** [comment]
Would it be strange to add the logs to cal_step instead of wfi_image?  That would make it more general, and possible relate any errors generated by cal better to the steps within the file.

**Comment by eslavich** [comment]
> Would it be strange to add the logs to cal_step instead of wfi_image?

We'd need to find a place to store the log messages not emitted by any particular step (from the pipeline itself, or the stpipe infrastructure).  Any ideas about that?

**Comment by sosey** [comment]
Related, what happens when a ramp model is saved after minimal processing? I think that would need to be updated too if we go with this approach.

I'm sure I'm missing something, but why can't all the  messages also be saved as part of cal_logs, all of them, external to each of the steps?

**Comment by eslavich** [comment]
> I'm sure I'm missing something, but why can't all the messages also be saved as part of cal_logs, all of them, external to each of the steps?

Oops, no reason, I was just confused.  Is this the right idea:

```yaml
meta:
  cal_step:
    assign_wcs: COMPLETE
    flat_field: COMPLETE
    dark: SKIPPED
    logs:
      - '2021-11-17T19:45:16.299Z :: stpipe.LoggingStep :: INFO :: Step LoggingStep running
        with args ().'
      - '2021-11-17T19:45:16.300Z :: stpipe.LoggingStep :: INFO :: Step LoggingStep parameters
        are: {''pre_hooks'': [], ''post_hooks'': [], ''output_file'': None, ''output_dir'':
        None, ''output_ext'': ''.asdf'', ''output_use_model'': False, ''output_use_index'':
        True, ''save_results'': False, ''skip'': False, ''suffix'': None, ''search_output_file'':
        True, ''input_dir'': ''''}'
      - '2021-11-17T19:45:16.300Z :: stpipe.LoggingStep :: WARNING :: Splines failed to
        reticulate'
```

**Comment by sosey** [comment]
> Oops, no reason, I was just confused. Is this the right idea:
> 
> ```yaml
> meta:
>   cal_step:
>     assign_wcs: COMPLETE
>     flat_field: COMPLETE
>     dark: SKIPPED
>     logs:
>       - '2021-11-17T19:45:16.299Z :: stpipe.LoggingStep :: INFO :: Step LoggingStep running
>         with args ().'
>       - '2021-11-17T19:45:16.300Z :: stpipe.LoggingStep :: INFO :: Step LoggingStep parameters
>         are: {''pre_hooks'': [], ''post_hooks'': [], ''output_file'': None, ''output_dir'':
>         None, ''output_ext'': ''.asdf'', ''output_use_model'': False, ''output_use_index'':
>         True, ''save_results'': False, ''skip'': False, ''suffix'': None, ''search_output_file'':
>         True, ''input_dir'': ''''}'
>       - '2021-11-17T19:45:16.300Z :: stpipe.LoggingStep :: WARNING :: Splines failed to
>         reticulate'
> ```

yup! that's what I was thinking.  What do you think?  Also,  @nden @perrygreenfield  might have opinions on this

**Comment by perrygreenfield** [comment]
I gave my opinion, except for where exactly the log would be located. No wait. I did. I had a preference for it to be at the end, and that is what Ed did (I figured leaving the verbose clutter at the end would make the rest of the yaml cleaner to view).

**Comment by PaulHuwe** [comment]
I also strongly prefer having the log at the end of the yaml header, since I (and future users) will want to sort through the yaml header to find information metadata keyword.

**Review by PaulHuwe — APPROVED** [review]

**Comment by PaulHuwe** [comment]
Have you performed a check of this code's effect on RDM? Should do that before merging.

**Comment by perrygreenfield** [comment]
@PaulHuwe Do you mean the data models?

**Comment by eslavich** [comment]
The corresponding roman_datamodels PR is here: https://github.com/spacetelescope/roman_datamodels/pull/53

**Comment by PaulHuwe** [comment]
> @PaulHuwe Do you mean the data models?

roman_Datamodels, yes.. and Ed already has the ticket - yay!

**Comment by sosey** [comment]
okay, so if we are moving it out of wfi_image and into cal_step, lets make sure the PR in roman_datamodels also gets the change

**Comment by eslavich** [comment]
(We chatted about this at the Periwinkle planning meeting and decided to keep the location of the logs as implemented)

**Pull request merged**